tests/base/generic/test_inputlist.py

```diff
@@ -0,0 +1,346 @@
+        pl.mark(2, "foo")
+        self.assertEqual(pl[2], pl.foo,
+                "marking with existing key broken")
+        with self.assertRaises(IndexError,
+                               msg = "no IndexError on invalid index"):
+            pl.mark(10, "foo")
+
+    def test_deep_copy(self):
+        pl = self.pl.copy()
+        self.assertTrue(pl is not self.pl,
+                "deep copy returns same object")
+        self.assertTrue(all(
+                pl[k1] is not self.pl[k2]
+                    for k1, k2 in zip(pl, self.pl)
+                        # int/str may be interned by python and always the same
+                        # object when equal, so exclude from the check
+                        if not isinstance(pl[k1], (int, str))),
+                "not a deep copy")
+        self.assertTrue(all(
+                (k1 == k2) and (pl[k1] == self.pl[k2])
+                    for k1, k2 in zip(pl, self.pl)),
+                "copy not equal to original")
+
+    def test_shallow_copy(self):
+        pl = copy.copy(self.pl)
+        self.assertTrue(pl is not self.pl,
+                "shallow copy returns same object")
+        self.assertTrue(all(
+                (k1 is k2) and (pl[k1] is self.pl[k2])
+                    for k1, k2 in zip(pl, self.pl)),
+                "not a shallow copy")
+        self.assertTrue(all(
+                (k1 == k2) and (pl[k1] == self.pl[k2])
+                    for k1, k2 in zip(pl, self.pl)),
+                "copy not equal to original")
+
+    def test_del_item(self):
+        pl = InputList({0: 1, "a": 2, "foo": 3})
+
+        with self.assertRaises(ValueError,
+                               msg = "no ValueError on invalid index type"):
+            del pl[{}]
+
+        del pl["a"]
+        self.assertTrue("a" not in pl, "delitem does not delete with str key")
+        del pl[0]
+        self.assertTrue(pl[0] != 1, "delitem does not delete with index")
+
+    def test_del_attr(self):
+        class SubInputList(InputList):
+            def __init__(self):
+                object.__setattr__(self, "attr", 42)
+        s = SubInputList()
+        del s.attr
+        self.assertFalse(hasattr(s, "attr"),
+                "delattr does not work with instance attributes")
+
+    def test_numpy_array(self):
+        pl = InputList([1,2,3])
+        self.assertTrue((np.array(pl) == np.array([1,2,3])).all(),
+                "conversion to numpy array broken")
+
+    def test_repr_json(self):
+        def rec(m):
+            """
+            Small helper to recurse through nested lists/dicts and check if all
+            keys and values are strings.  This should be the output format of
+            _repr_json_
+            """
+
+            if isinstance(m, list):
+                for v in m:
+                    if isinstance(v, (list, dict)):
+                        if not rec(v):
+                            return False
+                    elif type(v) != str:
```

Comment:
Is there a specific reason to not use `isinstance` here?

Reply:
My thinking was since the docstring promises `str` and there's no (reasonable) way `repr()` returns anything but a `str`, that's what should get checked.  I'm not married to it, though.

Reply:
`not isinstance(v, str):` ? I am just against using `type` and prefer using `isinstance` https://premaseem.wordpress.com/2017/10/11/python-difference-between-type-and-isinstance/